fix(pip_requirements): ignore extras in test for hashes

chludwig-haufe opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

Make the check in the `pip_requirements` manager whether a line contains a dependency with hashes ignore any extras between the dependency name and the equal signs.

## Context

Closes #16869

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

PhilipAbed commented on lib/modules/manager/pip_requirements/artifacts.ts:
```diff
@@ -25,10 +28,13 @@ export async function updateArtifacts({
       .split(newlineRegex)
       .map((line) => line.trim());
     for (const dep of updatedDeps) {
+      const depName = dep.depName!;
```
```suggestion
      if(!dep.Name) {continue;}
      const depName = dep.depName;
```

chludwig-haufe replied:
Fine with me. The existing code base used the non-null assertion operator, that's why I kept it.

chludwig-haufe replied:
Well, almost. Line 31 must read `if(!dep.depName) {continue;}`. :smile:

chludwig-haufe replied:
resolved in d540d0cf1fdb3e2621920edc50790f38ad65150e

PhilipAbed reviewed (commented):
looks good, thanks for the contribution!

PhilipAbed commented on lib/modules/manager/pip_requirements/artifacts.ts:
```diff
@@ -25,10 +28,13 @@ export async function updateArtifacts({
-          line.startsWith(`${dep.depName!}==`) && line.includes('--hash=')
+          line.startsWith(depName) &&
+          extrasAndEqualsRegex.test(line.substring(depName.length)) &&
+          line.includes('--hash=')
```
instead of these 3 lines, can we create 1 regex that supports all lines?

chludwig-haufe replied:
The `line.includes('--hash=')` won't be hard to cover by some regex. The dependency name is more problematic – it's a string we do not control and we either need to be 100% sure it does not contain any special characters or we need to escape it in the regex.

I am everything but a JavaScript or TypeScript expert. AFAIK – which boils down to the [regex chapter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions) in the MSDN JavaScript guide – the JavaScript regex engine does not support the `\Q`/`\E` quotation escapes or a function like Python's `re.escape()`.  Is there some helper function in the renovate codebase that escapes a string so it matches exactly the input string and nothing else?

Besides, keeping the test for the dependency name separate means the regular expression needs to be compiled only once, not for every single dependency. I don't expect this will have a huge impact in the end, though.

Regarding `lines.find( (line) => ...`:  I don't think it is needed in a strict sense. We certainly could come up with a regex that also accounts for the whitespace [trimmed from each line](https://github.com/renovatebot/renovate/blob/0b0fbd6ee81965ad652c69b1ff9f80fc759e010f/lib/modules/manager/pip_requirements/artifacts.ts#L29) and make sure the regex matches within single lines only. I am not sure whether the resulting code would be easier to understand, though. (The majority of the developers I know consider regular expressions hard to read.) And, of course, I did not want to make more changes to the codebase than necessary.

chludwig-haufe replied:
resolved in d540d0cf1fdb3e2621920edc50790f38ad65150e

chludwig-haufe replied:
Ok, I found `util.regex.escapeRegExp()`. I will give the single regex test a try.

chludwig-haufe replied:
It's in e8eccbf48099da83b0b6cc15b441160e363fc65e. You can decide whether you prefer the previous version or this one that requires the reader to think a bit more about the regexp.

PhilipAbed commented on lib/modules/manager/pip_requirements/artifacts.ts:
```diff
@@ -25,10 +28,13 @@ export async function updateArtifacts({
+          line.includes('--hash=')
       );
       if (hashLine) {
         const depConstraint = hashLine.split(' ')[0];
```
then here we wouldnt just split, instead you can have capture naming groups and put the group name here to understand what we are really picking up

chludwig-haufe replied:
Understood. All I need to know is how to proceed with escaping the dependency name inside the regex. Then I can push a change that extracts all the necessary bits during the regex matching.

PhilipAbed replied:
There is a function to escape depname its used on depnames all over the code ill check it tomorrow

chludwig-haufe replied:
I already found the escape function. e8eccbf48099da83b0b6cc15b441160e363fc65e uses a named subgroup to extract the dependency constraint.

PhilipAbed replied:
yup that's the one.

PhilipAbed commented on lib/modules/manager/pip_requirements/artifacts.ts:
```diff
@@ -4,8 +4,34 @@ import { logger } from '../../../logger';
+  // Use a non-greedy wildcard for the range pattern; otherwise, we would
+  // include all but the last hash specifier into depConstraint.
+  return new RegExp(
+    `^\\s*(?<depConstraint>${escapedDepName}${extrasPattern}\\s*==.*?\\S)\\s+--hash=`,
```
we usually use `regex` instead of the actual RegExp all over the code, 
the regex functionality takes 2 types of inputs
either regex('your regex', 'flag')
or regex(`/your regex/flag`) 
maybe more check it out 

i always had trouble with the 'm' flag which means Multiline right? let me know if it works for you
i'm not sure why, but when i do the 'm' flag it will not split lines correctly for me, and will not catch anything, 
so i use the regex on the whole content as one big string, and if i want another row id just add `\n` to the regex

chludwig-haufe replied:
No problem, it's changed. Should I set the parameter `useCache` to `false`? The expression will be different for every dependency. There might only be a small advantage to caching if the same dependency is used in multiple requirement files. (No idea whether they will then be processed within the same process, though...)

Yes, the flag `m` enables multiline-mode. (According to MDN, at least. :smile:) I updated the unit test so that the expression needs to match in the middle expression actually needs to match in the middle of the new package file content - the test passes. But, of course, I can also start the expression with `(?:^|\\n)` instead if you think it's more portable.

Regarding the test, I saw that the merge conflict will require some changes. I need to take a look at `test.exec-util.mockExecAll` before I can try to resolve the  conflict.

PhilipAbed replied:
As long as it works im fine with it xD test it on a real repo just to make sure.. looks good. U are doing great..

viceice replied:
Use our regEx function and allow to cache (it's in memory only, shouldn't matter much).

- The `m` specifier makes `^`  to match start of line and `$` to match end of line
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#advanced_searching_with_flags

so `^` is correct

chludwig-haufe replied:
The unit tests cover both cases (dependency at the beginning of the package file and in the middle) and I ran a successful test on [chrludwig/renovate-pip-requirements-extras-repro](https://github.com/chrludwig/renovate-pip-requirements-extras-repro).

viceice commented on lib/modules/manager/pip_requirements/artifacts.spec.ts:
```diff
@@ -1,17 +1,28 @@
 jest.mock('child_process');
 jest.mock('../../../util/exec');
 jest.mock('../../../util/fs');
+const mockedExec = mockedFunction(exec);
```
remove, just call mockExecAll in test and match object after artifacts call. checkout

viceice replied:
A good sample for using our exec mock

https://github.com/renovatebot/renovate/blob/1111b322f59aae97e5c54a550e5003946cf8bce9/lib/modules/manager/flux/artifacts.spec.ts#L62-L79

chludwig-haufe replied:
Thanks. And after reading the implementation of `mockExecAll` I understood what's happening.

In the end, it was most simple to accept the incoming `artifacts.spec.ts` during the merge "as is" and add my extra test cases in fe9d9161380cdf97c85eb6f7294897741e2dc061 again.

viceice commented on lib/modules/manager/pip_requirements/artifacts.ts:
```diff
@@ -21,17 +47,17 @@ export async function updateArtifacts({
+        // If there's a match, then the regular expression guarantees
+        // that the named subgroup deepConstraint did match as well.
+        const depConstraint = depAndHashMatch.groups!.depConstraint;
         cmd.push(`hashin ${depConstraint} -r ${packageFileName}`);
```
```suggestion
        cmd.push(`hashin ${quote(depConstraint)} -r ${quote(packageFileName)}`);
```

needs import update
```ts
import { quote } from 'shlex';
```

chludwig-haufe replied:
Done

viceice commented on lib/modules/manager/pip_requirements/artifacts.spec.ts:
```diff
@@ -54,53 +65,81 @@ describe('modules/manager/pip_requirements/artifacts', () => {
-          type: 'addition',
-          path: 'requirements.txt',
-          contents: 'new content',
+  it.each([
+    ['dependency w/o extras', 'atomicwrites', 'atomicwrites==1.4.0'],
+    [
+      'dependency with extras',
+      'boto3-stubs',
+      'boto3-stubs[iam] == 1.24.36.post1',
+    ],
+  ])(
```
please use a single test and pass all values as `updatedDeps`. So that `updateArtifacts` should update all hashes , so you get three exec snapshots

chludwig-haufe replied:
Now you say it... Makes sense, of course. Done.

chludwig-haufe commented:
I added a test that brings the test line coverage back to 100%.

Just as as side note (and maybe you are aware of this): The test `util/git/index › commitFiles({branchName, files, message}) › updates git submodules` kept failing on my machine (with the effect that the test coverage was always reported as < 100%) even though I had not touched any of the related files. In the end, I found the failure was somehow caused by the fact that I had a pre-commit hook [configured in the git init templates](https://pre-commit.com/#pre-commit-init-templatedir). Once I moved the hook template aside, the test succeeded locally as well. I did not look further into it, but maybe it's worth mentioning in your [local development docs](docs/development/local-development.md) that pre-commit templates can interfere with the tests.
